[Bug]: fails to compile: ‘AS_FORMAT_STYLE_COLLECTION’ undeclared

### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a bug that matches the one I want to file, without success.
- [X] If this is an issue with a particular app, I have tried filing it in the appropriate issue tracker for the app (e.g. under https://github.com/flathub/) and determined that it is an issue with Flatpak itself.
- [X] This issue is not a report of a security vulnerability (see [here](https://github.com/flatpak/flatpak/blob/main/SECURITY.md) if you need to report a security issue).

### Flatpak version

git master

### What Linux distribution are you using?

Fedora Linux

### Linux distribution version

39

### What architecture are you using?

x86_64

### How to reproduce

Attempt to compile flatpak.

```
main❯ ninja
[91/281] Compiling C object app/liblibflatpak-app.a.p/flatpak-builtins-utils.c.o
FAILED: app/liblibflatpak-app.a.p/flatpak-builtins-utils.c.o 
ccache cc -Iapp/liblibflatpak-app.a.p -Iapp -I../app -I. -I.. -Icommon -I../common -Isubprojects/libglnx -I../subprojects/libglnx -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/sysprof-4 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/gio-unix-2.0 -I/usr/include/appstream -I/usr/include/json-glib-1.0 -I/usr/include/ostree-1 -I/usr/include/polkit-1 -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -O0 -g -include config.h -Werror=shadow -Werror=empty-body -Werror=strict-prototypes -Werror=missing-prototypes -Werror=implicit-function-declaration -Werror=pointer-arith -Werror=init-self -Werror=missing-declarations -Werror=return-type -Werror=overflow -Werror=int-conversion -Werror=incompatible-pointer-types -Werror=misleading-indentation -Werror=missing-include-dirs -Wno-sign-compare -Wno-error=sign-compare -Wno-cast-function-type -Wno-error=cast-function-type -Wno-missing-field-initializers -Wno-error=missing-field-initializers -Wno-unused-parameter -Wno-error=unused-parameter -fvisibility=hidden -fPIC -pthread -MD -MQ app/liblibflatpak-app.a.p/flatpak-builtins-utils.c.o -MF app/liblibflatpak-app.a.p/flatpak-builtins-utils.c.o.d -o app/liblibflatpak-app.a.p/flatpak-builtins-utils.c.o -c ../app/flatpak-builtins-utils.c
../app/flatpak-builtins-utils.c: In function ‘flatpak_dir_load_appstream_store’:
../app/flatpak-builtins-utils.c:1133:40: error: ‘AS_FORMAT_STYLE_COLLECTION’ undeclared (first use in this function); did you mean ‘AS_FORMAT_STYLE_LAST’?
 1133 |   as_metadata_set_format_style (mdata, AS_FORMAT_STYLE_COLLECTION);
      |                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                        AS_FORMAT_STYLE_LAST
../app/flatpak-builtins-utils.c:1133:40: note: each undeclared identifier is reported only once for each function it appears in
[104/281] Compiling C object common/libflatpak-common.a.p/flatpak-dir.c.o
ninja: build stopped: subcommand failed.
```

### Expected Behavior

Compile

### Actual Behavior

Error

### Additional Information

_No response_

--- a/app/flatpak-builtins-utils.c
+++ b/app/flatpak-builtins-utils.c
@@ -1130,7 +1130,12 @@ flatpak_dir_load_appstream_store (FlatpakDir   *self,
                                        NULL);
 
   appstream_file = g_file_new_for_path (appstream_path);
+#if AS_CHECK_VERSION(0, 16, 0)
+  as_metadata_set_format_style (mdata, AS_FORMAT_STYLE_CATALOG);
+#else
+  /* Deprecated name for the same thing */
   as_metadata_set_format_style (mdata, AS_FORMAT_STYLE_COLLECTION);
+#endif
 #if AS_CHECK_VERSION(0, 14, 0)
   success = as_metadata_parse_file (mdata, appstream_file, AS_FORMAT_KIND_XML, &local_error);
 #else

--- a/app/flatpak-builtins-utils.c
+++ b/app/flatpak-builtins-utils.c
@@ -1131,7 +1131,7 @@ flatpak_dir_load_appstream_store (FlatpakDir   *self,
 
   appstream_file = g_file_new_for_path (appstream_path);
   as_metadata_set_format_style (mdata, AS_FORMAT_STYLE_COLLECTION);
-#ifdef HAVE_APPSTREAM_0_14_0
+#if AS_CHECK_VERSION(0, 14, 0)
   success = as_metadata_parse_file (mdata, appstream_file, AS_FORMAT_KIND_XML, &local_error);
 #else
   as_metadata_parse_file (mdata, appstream_file, AS_FORMAT_KIND_XML, &local_error);

--- a/app/flatpak-builtins-utils.h
+++ b/app/flatpak-builtins-utils.h
@@ -28,6 +28,12 @@
 #include "flatpak-dir-private.h"
 #include "flatpak-permission-dbus-generated.h"
 
+/* AS_CHECK_VERSION was introduced in 0.14.0; we still support 0.12.0, so
+ * behave as though versions without this macro are arbitrarily old */
+#ifndef AS_CHECK_VERSION
+#define AS_CHECK_VERSION(major, minor, micro) (0)
+#endif
+
 /* Appstream data expires after a day */
 #define FLATPAK_APPSTREAM_TTL 86400
 

--- a/configure.ac
+++ b/configure.ac
@@ -362,9 +362,6 @@ PKG_CHECK_MODULES(OSTREE, [ostree-1 >= $OSTREE_REQS])
 PKG_CHECK_MODULES(JSON, [json-glib-1.0])
 
 PKG_CHECK_MODULES(APPSTREAM, [appstream >= 0.12.0])
-PKG_CHECK_MODULES(APPSTREAM_0_14_0, appstream >= 0.14.0,
-                  [AC_DEFINE([HAVE_APPSTREAM_0_14_0], [1], [Define if appstream >= 0.14.0 is available])],
-                  [true])
 
 PKG_CHECK_MODULES(GDK_PIXBUF, [gdk-pixbuf-2.0])
 

--- a/meson.build
+++ b/meson.build
@@ -363,10 +363,6 @@ if glib_dep.version().version_compare('>=2.60')
   cdata.set('GLIB_VERSION_MIN_REQUIRED', 'GLIB_VERSION_2_60')
 endif
 
-if appstream_dep.version().version_compare('>=0.14.0')
-  cdata.set('HAVE_APPSTREAM_0_14_0', 1)
-endif
-
 if dconf_dep.found()
   cdata.set('HAVE_DCONF', 1)
 endif

